src/main/scala/beam/agentsim/agents/modalBehaviors/DrivesVehicle.scala

```diff
@@ -0,0 +1,186 @@
+package beam.agentsim.agents.modalBehaviors
+
+import beam.agentsim.agents.BeamAgent.BeamAgentData
+import beam.agentsim.agents.PersonAgent._
+import beam.agentsim.agents.modalBehaviors.DrivesVehicle.{EndLegTrigger, NotifyLegEnd, NotifyLegStart, StartLegTrigger}
+import beam.agentsim.agents.util.{AggregatorFactory, MultipleAggregationResult}
+import beam.agentsim.agents.vehicles.BeamVehicle.{AlightingConfirmation, BeamVehicleIdAndRef, BecomeDriver, BecomeDriverSuccess, BoardingConfirmation, UnbecomeDriver, UpdateTrajectory}
+import beam.agentsim.agents.vehicles.PassengerSchedule
+import beam.agentsim.agents.vehicles.{BeamVehicle, VehicleData}
+import beam.agentsim.agents.{BeamAgent, PersonAgent, TriggerShortcuts}
+import beam.agentsim.events.resources.vehicle._
+import beam.agentsim.scheduler.{Trigger, TriggerWithId}
+import beam.router.RoutingModel.BeamLeg
+import beam.sim.HasServices
+import org.matsim.api.core.v01.Id
+import org.matsim.vehicles.Vehicle
+
+import scala.collection.mutable.HashSet
+
+
+/**
+  * @author dserdiuk on 7/29/17.
+  */
+object DrivesVehicle {
+  case class StartLegTrigger(tick: Double, beamLeg: BeamLeg) extends Trigger
+  case class EndLegTrigger(tick: Double, beamLeg: BeamLeg) extends Trigger
+  case class NotifyLegEnd(tick: Double)
+  case class NotifyLegStart(tick: Double)
+}
+
+trait DrivesVehicle[T <: BeamAgentData] extends  TriggerShortcuts with HasServices with AggregatorFactory {
+  this: BeamAgent[T] =>
+
+  //TODO: double check that mutability here is legit espeically with the schedules passed in
+  protected var passengerSchedule: PassengerSchedule = PassengerSchedule()
+
+  protected var _currentTriggerId: Option[Long] = None
+  protected var _currentTick: Option[Double] = None
+  protected var _currentLeg: Option[BeamLeg] = None
+  protected var _currentVehicle: Option[BeamVehicleIdAndRef] = None
+  protected val _awaitingBoardConfirmation: HashSet[Id[Vehicle]] = HashSet()
```

Comment:
Correct me if I'm wrong.. So these hashsets will contain ids of vehicle ( including humanbodyvehicle)  only of the passengers that's is about to board/alight rather than all passengers during trip ? My point is there is no way we could end up with id replacement in those hashsets